### ISSUE TITLE
`use_package()` and `use_package_doc()` confusion

### ISSUE THREAD

**Issue by asadow** [r-lib/usethis#1759]
While reading R Packages 2e, I struggled a lot with understanding when to use these two functions.

For functions that are called often and one wants to avoid using the `::` to access a namescape, the book recommends to call ` use_package_doc()` and add `@import dplyr`, for example.

The similarity in names is a bit strange.

I'm not sure what to recommend. After years of using R, but weeks of developing packages, I found I struggled a lot with the concepts these functions and the book are trying to communicate. It feels like these aren't written with internal, business-oriented package development in mind (where e.g. dplyr will often be called). I.e. `use_package()` seems biased toward packages wherein packages like `stringr`, `dplyr`, `glue`, `lubridate`, etc. are rarely called.

**Comment by malcolmbarrett** [comment]
I think this might be a better comment for the book than for usethis. You only use `use_package_doc()` once to create package-level documentation. You'll use `use_package()` for most packages more often, so I think this distinction will clarify as you progress.

> It feels like these aren't written with internal, business-oriented package development in mind

Having worked on internal packages in a lot of settings, I disagree. Use `use_package()` for most situations; use `@importFrom` (and `usethis::use_import_from()`) when it's really helpful; use `@import` rarely when it's of great benefit (for instance, usethis `@import`s fs and rlang because they are critical infrastructure to the package). 

In other words, embrace `use_package()` and `pkg::fun()` for most applications, even if that's not how you would have written a script. Cultivate this mental model: for package development, the equivalent of 

```r
library(dplyr)
select(...)
```

is 

```r
use_package("dplyr")
```

```r
dplyr::select(...)
```

**Comment by jennybc** [comment]
Overall, I endorse @malcolmbarrett's response, so thanks for that.

And, yes, any further discussion would be more natural as an issue on the book, not usethis. But I suspect there's not really more to discuss.

> For functions that are called often and one wants to avoid using the :: to access a namescape, the book recommends to call  `use_package_doc()` and add `@import dplyr`, for example.

This is not what `use_package_doc()` is for and if that appears in the book, please open an issue, because it would be a mistake we should correct. This should always be a recommendation to call `use_package()` and, as @malcolmbarrett says, then make `::`- style calls.

**Comment by asadow** [comment]
@malcolmbarrett 

Thanks for this advice. So if my package uses a lot of `dplyr` (and thus it would be really helpful to not have function code cluttered with `dplyr::`), I should use a lot of `@importFrom`. I have the feeling you are suggesting that the use of `dplyr` generally does not count as "critical infrastructure".

I admit I find the `select()` example strange. I use it so often I'm failing to see the use in `dplyr::`. I don't see myself naming a function nor importing another package's function that collides with a function I often use, like `select()`.

**Comment by malcolmbarrett** [comment]
Explicit namespacing (`pkg::fun()`) is definitely the norm for package development. Personally, I think it's beneficial to stick with that norm for a little while you're learning development, because 1) that's what others will expect in package code 2) it helps others know exactly where functions come from. I can say I've used this approach with dplyr in internal packages quite a lot. 

That said, it's a judgment call. If you think this through and decide that's the best approach for you and your team, go for it. You still need to use `use_package()`, though, because that establishes the dependency in the first place. And make sure you run `check()` so that you know you're doing it all correctly. (As an aside, even when you're not submitting to CRAN, as in an internal package, it's still really useful to make it "CRAN-ready" because tools like `check()` usually make your package better in general)
